src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -231,21 +237,270 @@ def _estimate_inplane_rotations(
+            (each pair of images induces two estimates).
+        :param Riis: A nx3x3 array of estimates of self-relative rotations.
+        :return: vijs, viis
+        """
+        order = self.order
+        n_img = self.n_img
+
+        nchoose2 = int(n_img * (n_img - 1) / 2)
+        assert (
+            len(Riis) == n_img
+        ), f"There must be one self-relative rotation per image. Got {len(Riis)} Riis."
+        assert (
+            len(Rijs) == nchoose2
+        ), f"There must be n-choose-2 relative rotations. Got {len(Rijs)}."
+
+        # Estimate viis from Riis. vii = 1/order * sum(Rii^s) for s = 0, 1, ..., order.
+        viis = np.zeros((n_img, 3, 3))
+        for i, Rii in enumerate(Riis):
+            viis[i] = np.mean(
+                [np.linalg.matrix_power(Rii, s) for s in np.arange(order)], axis=0
+            )
+
+        # Estimate vijs via local handedness synchronization.
+        vijs = np.zeros((nchoose2, 3, 3))
+        e1 = [1, 0, 0]
+        opts = np.zeros((8, 3, 3))
+        scores_rank1 = np.zeros(8)
+        min_idxs = np.zeros((nchoose2, 3, 3))
+        pairs = all_pairs(n_img)
+        for idx, (i, j) in enumerate(pairs):
+            Rii = Riis[i]
+            Rjj = Riis[j]
+            Rij = Rijs[idx]
+
+            Rii_J = J_conjugate(Rii)
+            Rjj_J = J_conjugate(Rjj)
+
+            # vij should be a singular matrix.
+            # We test 8 combinations of handedness and rotation by {g, g^n-1} for singularity to determine:
+            # a. whether to transpose Rii
+            # b. whether to J-conjugate Rii
+            # c. whether to J-conjugate Rjj
+            if order == 3:
+                opts[0] = Rij + (Rii @ Rij @ Rjj) + (Rii.T @ Rij @ Rjj.T)
+                opts[1] = Rij + (Rii_J @ Rij @ Rjj) + (Rii_J.T @ Rij @ Rjj.T)
+                opts[2] = Rij + (Rii @ Rij @ Rjj_J) + (Rii.T @ Rij @ Rjj_J.T)
+                opts[3] = Rij + (Rii_J @ Rij @ Rjj_J) + (Rii_J.T @ Rij @ Rjj_J.T)
+
+                opts[4] = Rij + (Rii.T @ Rij @ Rjj) + (Rii @ Rij @ Rjj.T)
+                opts[5] = Rij + (Rii_J.T @ Rij @ Rjj) + (Rii_J @ Rij @ Rjj.T)
+                opts[6] = Rij + (Rii.T @ Rij @ Rjj_J) + (Rii @ Rij @ Rjj_J.T)
+                opts[7] = Rij + (Rii_J.T @ Rij @ Rjj_J) + (Rii_J @ Rij @ Rjj_J.T)
+
+                # Normalize
+                opts = opts / 3
+
+            else:
+                opts[0] = Rij + (Rii @ Rij @ Rjj)
+                opts[1] = Rij + (Rii_J @ Rij @ Rjj)
+                opts[2] = Rij + (Rii @ Rij @ Rjj_J)
+                opts[3] = Rij + (Rii_J @ Rij @ Rjj_J)
+
+                opts[4] = Rij + (Rii.T @ Rij @ Rjj)
+                opts[5] = Rij + (Rii_J.T @ Rij @ Rjj)
+                opts[6] = Rij + (Rii.T @ Rij @ Rjj_J)
+                opts[7] = Rij + (Rii_J.T @ Rij @ Rjj_J)
+
+                # Normalize
+                opts = opts / 2
+
+            for k, opt in enumerate(opts):
```

Comment:
Can remove loop since `svd` is vectorized and use `anorm`:
```
svals = svd(opt, compute_uv=False)
scores_rank1 = anorm(svals - e1, 1)
```

Reply:
Ah yes, thank you. Done.